sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteBundlesToFiles.java

```diff
@@ -121,24 +141,43 @@ public void processElement(ProcessContext c) throws Exception {
     String tempFilePrefix = resolveTempLocation(
         c.getPipelineOptions().getTempLocation(), "BigQueryWriteTemp", stepUuid);
     TableRowWriter writer = writers.get(c.element().getKey());
+    if (writer != null && writer.getByteSize() > Write.MAX_FILE_SIZE) {
+      // File is too big. Close it and open a new file.
+      TableRowWriter.Result result = writer.close();
+      c.output(new Result<>(result.resourceId.toString(), result.byteSize, c.element().getKey()));
+      writer = null;
+    }
     if (writer == null) {
-      writer = new TableRowWriter(tempFilePrefix);
-      writer.open(UUID.randomUUID().toString());
-      writers.put(c.element().getKey(), writer);
-      LOG.debug("Done opening writer {}", writer);
+      // Only create a new writer if we have fewer than maxNumWritersPerBundle already in thi
+      // bundle.
+      if (writers.size() <= maxNumWritersPerBundle) {
+        writer = new TableRowWriter(tempFilePrefix);
+        writer.open(UUID.randomUUID().toString());
+        writers.put(c.element().getKey(), writer);
+        LOG.debug("Done opening writer {}", writer);
+      }
```

Comment:
Perhaps put lines 174 and onwards into the "else" to the current "if", and remove the condition at line 160?